File: benchmarks/reference/go/spectral_norm.go
package main

import (
	"fmt"
	"math"
	"time"
)

func evalA(i, j int) float64 {
	ij := i + j
	return 1.0 / float64(ij*(ij+1)/2+i+1)
}

func evalATimesU(u, au []float64, n int) {
	for i := 0; i < n; i++ {
		sum := 0.0
		for j := 0; j < n; j++ {
			sum += evalA(i, j) * u[j]
		}
		au[i] = sum
	}
}

func evalAtTimesU(u, atu []float64, n int) {
	for i := 0; i < n; i++ {
		sum := 0.0
		for j := 0; j < n; j++ {
			sum += evalA(j, i) * u[j]
		}
		atu[i] = sum
	}
}

func evalAtaTimesU(u, atau, tmp []float64, n int) {
	evalATimesU(u, tmp, n)
	evalAtTimesU(tmp, atau, n)
}

func main() {
	n := 500
	u := make([]float64, n)
	v := make([]float64, n)
	tmp := make([]float64, n)

	for i := range u {
		u[i] = 1.0
	}

	start := time.Now()

	for i := 0; i < 10; i++ {
		evalAtaTimesU(u, v, tmp, n)
		evalAtaTimesU(v, u, tmp, n)
	}

	vbv, vv := 0.0, 0.0
	for i := 0; i < n; i++ {
		vbv += u[i] * v[i]
		vv += v[i] * v[i]
	}

	ms := time.Since(start).Milliseconds()
	fmt.Printf("spectral norm: %f\n", math.Sqrt(vbv/vv))
	fmt.Printf("elapsed: %d ms\n", ms)
}

File: benchmarks/reference/python/spectral_norm.py
import math
import time

def eval_a(i, j):
    ij = i + j
    return 1.0 / (ij * (ij + 1) // 2 + i + 1)

def eval_a_times_u(u, n):
    au = [0.0] * n
    for i in range(n):
        s = 0.0
        for j in range(n):
            s += eval_a(i, j) * u[j]
        au[i] = s
    return au

def eval_at_times_u(u, n):
    atu = [0.0] * n
    for i in range(n):
        s = 0.0
        for j in range(n):
            s += eval_a(j, i) * u[j]
        atu[i] = s
    return atu

def eval_ata_times_u(u, n):
    return eval_at_times_u(eval_a_times_u(u, n), n)

if __name__ == "__main__":
    n = 500
    u = [1.0] * n

    t0 = time.monotonic_ns()

    for _ in range(10):
        v = eval_ata_times_u(u, n)
        u = eval_ata_times_u(v, n)

    vbv = sum(u[i] * v[i] for i in range(n))
    vv = sum(v[i] * v[i] for i in range(n))

    ms = (time.monotonic_ns() - t0) // 1_000_000
    print(f"spectral norm: {math.sqrt(vbv / vv):.6f}")
    print(f"elapsed: {ms} ms")
